hosting/CONFIGURATION.md
## Environment Variables

When running a Plausible release, the following configuration parameters can be supplied as environment variables.

### Default User Generation

For self-hosting, a default user can be generated using the `db init-admin` command. In the default docker-compose setup, this command
is run automatically when you boot the server.

It is *highly* recommended that you configure these parameters.

| Parameter        | Default       | Description                          |
|------------------|---------------|--------------------------------------|
| ADMIN_USER_NAME  | autogenerated | Admin user's name                    |
| ADMIN_USER_EMAIL | autogenerated | The default ("admin") user email     |
| ADMIN_USER_PWD   | autogenerated | The default ("admin") user password. |

### Server

Following are the variables that can be used to configure the availability of the server.

| Parameter            | Default     | Description                                                                                                                                                                                         |
|----------------------|-------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| HOST                 | localhost   | The hosting address of the server. For running on local system, this can be set to **localhost**. In production systems, this can be your ingress host.                                             |
| SCHEME               | http        | The scheme of the URL, either `http` or `https`. When using a reverse proxy with https, it'll be required to set this                                                                               |
| PORT                 | 8000        | The port on which the server is available.                                                                                                                                                          |
| SECRET_KEY_BASE      | --          | An internal secret key used by [Phoenix Framework](https://www.phoenixframework.org/). Follow the [instructions](https://hexdocs.pm/phoenix/Mix.Tasks.Phx.Gen.Secret.html#content) to generate one. |
| DISABLE_AUTH         | false       | Disables authentication completely, no registration, login will be shown.                                                                                                                           |
| DISABLE_REGISTRATION | false       | Disables registration of new users, keep your admin credentials handy ;)                                                                                                                            |

### Database

Plausible uses [PostgreSQL as database](https://www.tutorialspoint.com/postgresql/postgresql_environment.htm) for storing all the user-data. Use the following variables to configure it.

| Parameter            | Default                                                   | Description                                                                                                               |
|----------------------|-----------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------|
| DATABASE_URL         | postgres://postgres:postgres@127.0.0.1:5432/plausible_dev | The database URL as dictated [here](https://hexdocs.pm/ecto/Ecto.Repo.html#module-urls)                                   |
| DATABASE_POOL_SIZE   | 10                                                        | Number of connections to the database                                                                                     |
| DATABASE_TLS_ENABLED | false                                                     | A flag that says whether to connect to the database via TLS, read [here](https://www.postgresql.org/docs/10/ssl-tcp.html) |

For performance reasons, all the analytics events are stored in [clickhouse](https://clickhouse.tech/docs/en/getting-started/tutorial/):

| Parameter                    | Default                             | Description                                                                                         |
|------------------------------|-------------------------------------|-----------------------------------------------------------------------------------------------------|
| CLICKHOUSE_DATABASE_URL      | http://localhost:8123/plausible_dev | Connection string for Clickhouse. The protocol is either `http` or `https` depending on your setup. |
| CLICKHOUSE_DATABASE_POOLSIZE | 10                                  | Number of connections to the database                                                               |


### Mailer/SMTP Setup

Plausible uses and SMTP server to send transactional emails e.g. account activation, password reset, weekly reports, etc.

| Parameter             | Default   | Description                                      |
|-----------------------|-----------|--------------------------------------------------|
| MAILER_EMAIL          | hello@plausible.local  | The email id to use for as _from_ address of all communications from Plausible.|
| SMTP_HOST_ADDR        | localhost              | The host address of your smtp server.                                          |
| SMTP_HOST_PORT        | 25                     | The port of your smtp server.                                                  |
| SMTP_USER_NAME        | --                     | The username/email in case SMTP auth is enabled.                               |
| SMTP_USER_PWD         | --                     | The password in case SMTP auth is enabled.                                     |
| SMTP_HOST_SSL_ENABLED | false                  | If SSL is enabled for SMTP connection                                          |
| SMTP_RETRIES          | 2                      | Number of retries to make until mailer gives up.                               |

Alternatively, you can use Postmark to send transactional emails. In this case, use the following parameters:

| Parameter          | Default            | Description                                                                                          |
|--------------------|--------------------|------------------------------------------------------------------------------------------------------|
| MAILER_ADAPTER     | Bamboo.SMTPAdapter | Instead of the default, replace this with `Bamboo.PostmarkAdapter`  |
| POSTMARK_API_KEY   | --                 | Enter your API key.                                                 |


### IP Geolocation

Plausible uses the GeoLite2 database created by [MaxMind](https://www.maxmind.com) for enriching analytics data with visitor countries. Their
end-user license does not make it very easy to just package the database along with an open-source product. This is why, if you want
to get country data for your analytics, you need to create an account and download their **GeoLite2 Country** database.

Once you have the database, mount it on the Plausible docker image and configure the path of the database file:

| Parameter           | Default   | Description                                                                                          |
|---------------------|-----------|------------------------------------------------------------------------------------------------------|
| GEOLITE2_COUNTRY_DB | --        | Path to your IP geolocation database in MaxMind's format  |

To make this as easy as possible you can use the [`maxmindinc/geoipupdate`](https://hub.docker.com/r/maxmindinc/geoipupdate) Docker image.
You just need to add your account details, mount the database in the `plausible` container and let the image update the database automatically.
To run the complete setup including geoip see [`docker-compose-geoip.yml`](docker-compose-geoip.yml#L35).

If the Geolite database is not configured, no country data will be captured.

### External Services

- [Google Client](https://developers.google.com/api-client-library)
    - GOOGLE_CLIENT_ID
    - GOOGLE_CLIENT_SECRET
- [Sentry](https://sentry.io/)
    - SENTRY_DSN
- [Paddle](https://paddle.com/)
    - PADDLE_VENDOR_AUTH_CODE
- [PostMark](https://postmarkapp.com/), only in case of `Bamboo.PostmarkAdapter` mail adapter.
    - POSTMARK_API_KEY

Apart from these, there are also the following integrations

- [Twitter](https://developer.twitter.com/en/docs)
    - TWITTER_CONSUMER_KEY
    - TWITTER_CONSUMER_SECRET
    - TWITTER_ACCESS_TOKEN
    - TWITTER_ACCESS_TOKEN_SECRET
- [Slack](https://api.slack.com/messaging/webhooks)
    - SLACK_WEBHOOK


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.0] - Unreleased

### Added
- Collect and present link tags (`utm_medium`, `utm_source`, `utm_campaign`) in the dashboard

### Changed
- Replace configuration parameters `CLICKHOUSE_DATABASE_{HOST,NAME,USER,PASSWORD}` with a single `CLICKHOUSE_DATABASE_URL` [plausible/analytics#317](https://github.com/plausible/analytics/pull/317)
- Disable subscriptions by default
- Remove `CLICKHOUSE_DATABASE_POOLSIZE`, `DATABASE_POOLSIZE` and `DATABASE_TLS_ENABLED` parameters. Use query parameters in `CLICKHOUSE_DATABASE_URL` and `DATABASE_URL` instead.
- Remove `HOST` and `SCHEME` parameters in favor of a single `BASE_URL` parameter.
- Make `Bamboo.SMTPAdapter` the default as opposed to `Bamboo.PostmarkAdapter`
- Disable subscription flow by default


config/config.exs
use Mix.Config

base_url = System.get_env("BASE_URL", "http://localhost:8000")
           |> URI.parse

config :plausible,
  admin_user: System.get_env("ADMIN_USER_NAME", "admin"),
  admin_email: System.get_env("ADMIN_USER_EMAIL", "admin@plausible.local"),
  mailer_email: System.get_env("MAILER_EMAIL", "hello@plausible.local"),
  admin_pwd: System.get_env("ADMIN_USER_PWD", "!@d3in"),
  ecto_repos: [Plausible.Repo, Plausible.ClickhouseRepo],
  environment: System.get_env("ENVIRONMENT", "dev")

disable_auth = String.to_existing_atom(System.get_env("DISABLE_AUTH", "false"))

config :plausible, :selfhost,
  disable_authentication: disable_auth,
  disable_subscription: String.to_existing_atom(System.get_env("DISABLE_SUBSCRIPTION", "false")),
  disable_registration:
    if(disable_auth,
      do: true,
      else: String.to_existing_atom(System.get_env("DISABLE_REGISTRATION", "false"))
    )

# Configures the endpoint
config :plausible, PlausibleWeb.Endpoint,
  url: [
    host: base_url.host,
    scheme: base_url.scheme,
    port: base_url.port
  ],
  http: [
    port: String.to_integer(System.get_env("PORT", "8000"))
  ],
  secret_key_base:
    System.get_env(
      "SECRET_KEY_BASE",
      "/NJrhNtbyCVAsTyvtk1ZYCwfm981Vpo/0XrVwjJvemDaKC/vsvBRevLwsc6u8RCg"
    ),
  render_errors: [
    view: PlausibleWeb.ErrorView,
    accepts: ~w(html json)
  ],
  pubsub: [name: Plausible.PubSub, adapter: Phoenix.PubSub.PG2]

config :sentry,
  dsn: System.get_env("SENTRY_DSN"),
  included_environments: [:prod, :staging],
  environment_name: String.to_atom(System.get_env("ENVIRONMENT", "dev")),
  enable_source_code_context: true,
  root_source_code_path: File.cwd!(),
  tags: %{app_version: System.get_env("APP_VERSION", "0.0.1")},
  context_lines: 5

# Configures Elixir's Logger
config :logger, :console,
  format: "$time $metadata[$level] $message\n",
  metadata: [:request_id]

# Use Jason for JSON parsing in Phoenix
config :phoenix, :json_library, Jason

config :ua_inspector,
  database_path: "priv/ua_inspector"

config :ref_inspector,
  database_path: "priv/ref_inspector"

config :plausible,
  paddle_api: Plausible.Billing.PaddleApi,
  google_api: Plausible.Google.Api

config :plausible,
  # 30 minutes
  session_timeout: 1000 * 60 * 30,
  session_length_minutes: 30

config :plausible, :paddle,
  vendor_id: "49430",
  vendor_auth_code: System.get_env("PADDLE_VENDOR_AUTH_CODE")

config :plausible, Plausible.ClickhouseRepo,
  loggers: [Ecto.LogEntry],
  url: System.get_env(
    "CLICKHOUSE_DATABASE_URL",
    "http://127.0.0.1:8123/plausible_test"
  )

config :plausible,
       Plausible.Repo,
       timeout: 300_000,
       connect_timeout: 300_000,
       handshake_timeout: 300_000,
       url:
         System.get_env(
           "DATABASE_URL",
           "postgres://postgres:postgres@127.0.0.1:5432/plausible_dev?currentSchema=default"
         )

cron_enabled = String.to_existing_atom(System.get_env("CRON_ENABLED", "false"))

base_cron = [
  # Daily at midnight
  {"0 0 * * *", Plausible.Workers.RotateSalts}
]

extra_cron = [
  # hourly
  {"0 * * * *", Plausible.Workers.SendSiteSetupEmails},
  #  hourly
  {"0 * * * *", Plausible.Workers.ScheduleEmailReports},
  # Daily at midnight
  {"0 0 * * *", Plausible.Workers.FetchTweets},
  # Daily at midday
  {"0 12 * * *", Plausible.Workers.SendTrialNotifications},
  # Daily at midday
  {"0 12 * * *", Plausible.Workers.SendCheckStatsEmails},
  # Every 10 minutes
  {"*/10 * * * *", Plausible.Workers.ProvisionSslCertificates}
]

base_queues = [rotate_salts: 1]

extra_queues = [
  provision_ssl_certificates: 1,
  fetch_tweets: 1,
  check_stats_emails: 1,
  site_setup_emails: 1,
  trial_notification_emails: 1,
  schedule_email_reports: 1,
  send_email_reports: 1
]

config :plausible, Oban,
  repo: Plausible.Repo,
  queues: if(cron_enabled, do: base_queues ++ extra_queues, else: base_queues),
  crontab: if(cron_enabled, do: base_cron ++ extra_cron, else: base_cron)

config :plausible, :google,
  client_id: System.get_env("GOOGLE_CLIENT_ID"),
  client_secret: System.get_env("GOOGLE_CLIENT_SECRET")

config :plausible, :slack, webhook: System.get_env("SLACK_WEBHOOK")

mailer_adapter = System.get_env("MAILER_ADAPTER", "Bamboo.LocalAdapter")

case mailer_adapter do
  "Bamboo.PostmarkAdapter" ->
    config :plausible, Plausible.Mailer,
      adapter: :"Elixir.#{mailer_adapter}",
      api_key: System.get_env("POSTMARK_API_KEY")

  "Bamboo.SMTPAdapter" ->
    config :plausible, Plausible.Mailer,
      adapter: :"Elixir.#{mailer_adapter}",
      server: System.fetch_env!("SMTP_HOST_ADDR"),
      hostname: System.get_env("HOST", "localhost"),
      port: System.fetch_env!("SMTP_HOST_PORT"),
      username: System.fetch_env!("SMTP_USER_NAME"),
      password: System.fetch_env!("SMTP_USER_PWD"),
      tls: :if_available,
      allowed_tls_versions: [:tlsv1, :"tlsv1.1", :"tlsv1.2"],
      ssl: System.get_env("SMTP_HOST_SSL_ENABLED") || false,
      retries: System.get_env("SMTP_RETRIES") || 2,
      no_mx_lookups: System.get_env("SMTP_MX_LOOKUPS_ENABLED") || true,
      auth: :if_available

  "Bamboo.LocalAdapter" ->
    config :plausible, Plausible.Mailer, adapter: :"Elixir.#{mailer_adapter}"

  _ ->
    raise "Unknown mailer_adapter; expected SMTPAdapter or PostmarkAdapter"
end

config :plausible, :twitter,
  consumer_key: System.get_env("TWITTER_CONSUMER_KEY"),
  consumer_secret: System.get_env("TWITTER_CONSUMER_SECRET"),
  token: System.get_env("TWITTER_ACCESS_TOKEN"),
  token_secret: System.get_env("TWITTER_ACCESS_TOKEN_SECRET")

config :plausible, :custom_domain_server,
  user: System.get_env("CUSTOM_DOMAIN_SERVER_USER"),
  password: System.get_env("CUSTOM_DOMAIN_SERVER_PASSWORD"),
  ip: System.get_env("CUSTOM_DOMAIN_SERVER_IP")

config :plausible, PlausibleWeb.Firewall,
  blocklist: System.get_env("IP_BLOCKLIST", "") |> String.split(",") |> Enum.map(&String.trim/1)

config :plausible, :hcaptcha,
  sitekey: System.get_env("HCAPTCHA_SITEKEY"),
  secret: System.get_env("HCAPTCHA_SECRET")

config :geolix,
  databases: [
    %{
      id: :country,
      adapter: Geolix.Adapter.MMDB2,
      source: "priv/geolix/GeoLite2-Country.mmdb"
    }
  ]

# Import environment specific config. This must remain at the bottom
# of this file so it overrides the configuration defined above.
import_config "#{Mix.env()}.exs"
